File: netCDF/hands_on/F90/hands_on_par_write.f90
program hands_on_par_write 
  use netcdf
  implicit none
  include 'mpif.h'

  integer :: rank
  integer :: mpisize
  integer :: ierr
  
  character (len = *), parameter :: FILE_NAME = "par_write.nc"
  integer,parameter :: NDIMS=2
  integer,parameter :: NI_GLO=5
  integer,parameter :: NJ_GLO=10
  integer,parameter :: NTIME=2

  real:: lon_glo(NJ_GLO),lat_glo(NI_GLO)
  integer :: field_glo(NI_GLO,NJ_GLO)
  real,allocatable :: lon(:),lat(:)
  integer,allocatable :: field(:,:) 
  integer :: ni,ibegin,nj,jbegin
  integer :: t
  integer :: ncid,lat_dimid,lon_dimid,timedimid,varid,lonid,latid

  integer :: dimids(NDIMS), starts(NDIMS), counts(NDIMS)

  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,rank,ierr)
  call MPI_Comm_size(MPI_COMM_WORLD,mpisize,ierr)

  ! Create the netcdf file.
  call check(...)
  
  ! Define three dimensions lat (size NI_GLO), lon (size NJ_GLO) and time. 
  call check(...)
  call check(...)
  call check(...) 

  ! Initialize global arrays.
  call init_global(lon_glo,lat_glo,field_glo)

  ! Decompose global domain. Returned values: ni,nj,ibegin,jbegin.
  call decompose_domain(mpisize,rank,ni,ibegin,nj,jbegin)

  ! Allocate and fill in local arrays. 
  allocate(lat(ni),lon(nj),field(ni,nj))
  lat(:)=lat_glo(ibegin:ibegin-1+ni)
  lon(:)=lon_glo(jbegin:jbegin-1+nj)
  field(:,:) = field_glo(ibegin:ibegin-1+ni,jbegin:jbegin-1+nj)

  ! Define variable "lat" containing latitude.
  call check(...)
  ! Add attribute "standart_name" holding the value "latitude"
  call check(...)
  ! Add attribute "units" holding the value "degrees_north"
  call check(...)

  ! Define variable "lon" containing longitude.
  call check(...)
  ! Add attribute "standart_name" holding the value "longitude"
  call check(...)
  ! Add attribute "units" holding the value "degrees_east"
  call check(...)

  ! Define variable "var_2d_temp".
  call check(...)
  ! Add attribute "standart_name" holding the value "dummy_temporal_variable"
  call check(...)

  ! End define mode. This tells netCDF we are done defining metadata.
  call check(...)

  ! Write latitude
  call check(...)
 ! Write longitude
  call check(...)

  ! Set mode to collective for the temporal variable.
  call check(...)) 

  ! Write the temporal variable.
  do t=1,NTIME
    call check(...)
  enddo

  ! Close file.
  call check(...)

  deallocate(lon, lat, field)
  call MPI_FINALIZE(ierr)

  contains
  ! This subroutine comes with the netCDF package. 
  ! It handles errors by printing an error message  and exiting with a non-zero status.
  subroutine check(status)
    integer,intent(in) :: status
    if(status /= nf90_noerr) then 
      print *, trim(nf90_strerror(status))
      stop "Stopped"
    end if
  end subroutine check

  ! This subroutine fills in global arrays containing values of the variable, longitude, and latitude.
  subroutine init_global(lon_glo,lat_glo,field_glo)
  real,intent(inout) :: lon_glo(NJ_GLO),lat_glo(NI_GLO)
  integer,intent(inout) :: field_glo(NI_GLO,NJ_GLO)        
  integer i,j
  
  do i=1,NI_GLO
    lat_glo(i) = -90.+180./NI_GLO*(i-0.5)
  enddo
  do j=1,NJ_GLO
    lon_glo(j) = -180.+360./NJ_GLO*(j-0.5)
    do i=1,NI_GLO
      field_glo(i,j)=(i-1)+(j-1)*NI_GLO
    enddo
  enddo
  end subroutine init_global

  ! This subroutine decomposes global domain.
  subroutine decompose_domain(mpisize,rank,ni,ibegin,nj,jbegin)
    integer,intent(in) :: mpisize,rank
    integer,intent(out) :: ni,ibegin,nj,jbegin
    integer :: n
    ni=NI_GLO ; ibegin=1
    jbegin=0
    do n=0,mpisize-1
      nj=NJ_GLO/mpisize
      if (n<MOD(NJ_GLO,mpisize)) nj=nj+1
      if (n==rank) exit
      jbegin=jbegin+nj
    enddo
    jbegin=jbegin+1
  end subroutine decompose_domain

end program hands_on_par_write

File: netCDF/hands_on/Solutions/C/hands_on_par_write.c
#include "mpi.h"
#include <netcdf.h>
#include <netcdf_par.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NDIMS     2
#define NI_GLO    5 
#define NJ_GLO    10
#define NTIME     2

#define ERR(e) {printf("Error: %s\n", nc_strerror(e)); return 2;}

// Function decomposes global domain of size NI_GLO*NJ_GLO in horizontal direction into local domains
void decompose_domain(int mpisize, int mpirank, int* ni, int* ibegin, int* nj, int* jbegin)
{
    *nj=NJ_GLO;
    *ibegin=0;
    *jbegin=0;
    for (int n=0; n<mpisize; ++n)
    {	    
      *ni=NI_GLO/mpisize;
      if (n<(NI_GLO%mpisize)) ++(*ni);
      if (n==mpirank) break;	      
      (*ibegin)+=(*ni);
    }  
}

// Function fills in local domain data
void fill_local_data(int* data, int ni, int ibegin, int nj, int jbegin)
{
  int i,j;
  for(i=0; i < ni; i++)
    for(j=0; j < nj; j++)
      data[j+i*nj] = (j+jbegin) + (i+ibegin)*nj;
}

// Function fills in local lon and lat
void fill_lon_lat(float* lon, float* lat, int ni, int ibegin, int nj, int jbegin)
{
  for(int i=0; i < ni; i++)
    lat[i] = -90. + 180./NI_GLO*(i+ibegin+0.5);
  for(int j=0; j < nj; j++)
    lon[j] = -180. + 360./NJ_GLO*(j+jbegin+0.5);
}

int main (void)
{
  int mpirank, mpisize;
  /* Variables holding netCDF ids of the file and three variables stored in it (longitude, latitude and a dummy 2D variable) */
  int ncid, lonid, latid, varid;
  /* Variables holding three dimensions: longitude, latitude and a temporal dimension  */
  int lon_dimid, lat_dimid, temp_dimid;
  /* Array holding three dimensions of the dummy 2D varialbe */
  int var_dimids[NDIMS+1];
  /* Start and count for parallel write of longitude, latitude and the variable */ 
  size_t start, count;
  size_t var_starts[NDIMS+1], var_counts[NDIMS+1];
  int err;
  int ni, nj, ibegin, jbegin;
  char filename[128] = "par_write.nc";

  //--------------------------------------
  // Initialization MPI
  //--------------------------------------

  MPI_Init(NULL, NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&mpirank);
  MPI_Comm_size(MPI_COMM_WORLD,&mpisize);

  //--------------------------------------
  //  NetCDF metadata definitions
  //--------------------------------------

  /* Create a netcdf file */
  if (err=nc_create_par(filename, NC_NETCDF4, MPI_COMM_WORLD, MPI_INFO_NULL, &ncid))
    ERR(err);

  /* Define two spatial dimensions */
  if (err=nc_def_dim(ncid, "lat", NI_GLO, &lat_dimid))
    ERR(err);
  if (err=nc_def_dim(ncid, "lon", NJ_GLO, &lon_dimid))
    ERR(err);

  /* Define a temporal dimension */
  if (err=nc_def_dim(ncid, "time", NC_UNLIMITED, &temp_dimid))
    ERR(err);

  /* Define 1D variables that will hold latitude and longitude */
  if (err=nc_def_var(ncid, "lat", NC_FLOAT, 1, &lat_dimid, &latid)) 
    ERR(err);
  if (err=nc_put_att(ncid, latid, "standart_name", NC_CHAR, strlen("latitude"), "latitude"))  
    ERR(err);
  if (err=nc_put_att(ncid, latid, "units", NC_CHAR, strlen("degrees_north"), "degrees_north"))  
    ERR(err);
  if (err=nc_def_var(ncid, "lon", NC_FLOAT, 1, &lon_dimid, &lonid))
    ERR(err);
  if (err=nc_put_att(ncid, lonid, "standart_name", NC_CHAR, strlen("longitude"), "longitude"))
    ERR(err);
  if (err=nc_put_att(ncid, lonid, "units", NC_CHAR, strlen("degrees_east"), "degrees_east"))
    ERR(err);

  /* Define a 2D temporal variable and add an attribute */
  var_dimids[0] = temp_dimid;
  var_dimids[1] = lat_dimid;
  var_dimids[2] = lon_dimid;
  if (err=nc_def_var(ncid, "var_2d_temp", NC_INT, NDIMS+1, var_dimids, &varid))
    ERR(err);
  if (err=nc_put_att(ncid, varid, "standart_name", NC_CHAR, strlen("dummy_temporal_variable"), "dummy_temporal_variable"))
    ERR(err);

  /* Close metadata definition */
  if (err=nc_enddef(ncid))
    ERR(err);

  //--------------------------------------
  // Local data initialization
  //--------------------------------------

  /* Decompose global domain. Returned values: ni,nj,ibegin,jbegin */
  decompose_domain(mpisize,mpirank,&ni,&ibegin,&nj,&jbegin);
  
  /* Fill in local data: lon, lat and temporal varialbe */
  int *data = (int *)malloc(sizeof(int)*ni*nj);
  fill_local_data(data, ni, ibegin, nj, jbegin);
  
  float *lon = (float *)malloc(sizeof(float)*nj);
  float *lat = (float *)malloc(sizeof(float)*ni);
  fill_lon_lat(lon, lat, ni, ibegin, nj, jbegin);

  //--------------------------------------
  // Write data and close the file  
  //--------------------------------------

  /* Write lon, lat.
   * Here, the third and the fourth arguments, start and count, should be of type 'size_t *'.*/
  start = ibegin; count = ni;
  if (err=nc_put_vara(ncid, latid, &start, &count, lat)) ERR(err);
  start = jbegin; count = nj;
  if (err=nc_put_vara(ncid, lonid, &start, &count, lon)) ERR(err);  

  var_starts[0] = 0;  var_starts[1] = ibegin;  var_starts[2] = jbegin;
  var_counts[0] = 1;  var_counts[1] = ni;      var_counts[2] = nj;

  /* Set mode to collective for temporal variable */
  if (err=nc_var_par_access(ncid, varid, NC_COLLECTIVE)) ERR(err);

  /* Write temporal data */
  for (int t=0; t<NTIME; ++t)
  { 
    var_starts[0] = t;
    if (err=nc_put_vara(ncid, varid, var_starts, var_counts, data)) ERR(err);
  }

  /* Close the file */
  if (err=nc_close(ncid)) ERR(err);

  free(data);
  free(lon);
  free(lat);

  MPI_Finalize();

  return 0;
}     

